cmd/kube-proxy/app/server.go

```diff
@@ -567,14 +568,117 @@ func newProxyServer(config *kubeproxyconfig.KubeProxyConfiguration, master strin
+	} else {
+		klog.InfoS("kube-proxy running in single-stack mode", "ipFamily", s.PrimaryIPFamily)
+	}
+
+	err, fatal := checkIPConfig(s, dualStackSupported)
+	if err != nil {
+		if fatal {
+			return nil, fmt.Errorf("kube-proxy configuration is incorrect: %v", err)
+		}
+		klog.ErrorS(err, "Kube-proxy configuration may be incomplete or incorrect")
+	}
+
+	s.Proxier, err = s.createProxier(config, dualStackSupported)
 	if err != nil {
 		return nil, err
 	}
 
 	return s, nil
 }
 
+// checkIPConfig confirms that s has proper configuration for its primary IP family.
+func checkIPConfig(s *ProxyServer, dualStackSupported bool) (error, bool) {
+	var errors []error
+	var okIP, okCIDR func(string) bool
+	var secondaryIPFamily v1.IPFamily
+
+	if s.PrimaryIPFamily == v1.IPv4Protocol {
+		secondaryIPFamily = v1.IPv6Protocol
+		okIP = netutils.IsIPv4String
+		okCIDR = netutils.IsIPv4CIDRString
+	} else {
+		secondaryIPFamily = v1.IPv4Protocol
+		okIP = netutils.IsIPv6String
+		okCIDR = netutils.IsIPv6CIDRString
+	}
+
+	var clusterType string
+	if dualStackSupported {
+		clusterType = fmt.Sprintf("%s-primary", s.PrimaryIPFamily)
+	} else {
+		clusterType = fmt.Sprintf("%s-only", s.PrimaryIPFamily)
+	}
+
+	// Historically, we did not check most of the config options, so we cannot
+	// retroactively make IP family mismatches in those options be fatal. When we add
+	// new options to check here, we should make problems with those options be fatal.
+	fatal := false
+
+	if s.Config.ClusterCIDR != "" {
+		clusterCIDRs := strings.Split(s.Config.ClusterCIDR, ",")
+		if !anyMatch(clusterCIDRs, okCIDR) {
+			errors = append(errors, fmt.Errorf("cluster is %s but clusterCIDRs contains only %s addresses", clusterType, secondaryIPFamily))
+		}
+	}
+
+	if len(s.Config.NodePortAddresses) > 0 && !anyMatch(s.Config.NodePortAddresses, okCIDR) {
+		errors = append(errors, fmt.Errorf("cluster is %s but nodePortAddresses contains only %s addresses", clusterType, secondaryIPFamily))
+	}
+
+	if len(s.podCIDRs) > 0 && !anyMatch(s.podCIDRs, okCIDR) {
+		errors = append(errors, fmt.Errorf("cluster is %s but node.spec.podCIDRs contains only %s addresses", clusterType, secondaryIPFamily))
+		if s.Config.DetectLocalMode == kubeproxyconfig.LocalModeNodeCIDR {
+			// This has always been a fatal error
+			fatal = true
+		}
+	}
+
+	if s.Config.Winkernel.SourceVip != "" && !okIP(s.Config.Winkernel.SourceVip) {
+		errors = append(errors, fmt.Errorf("cluster is %s but winkernel.sourceVip is %s", clusterType, secondaryIPFamily))
+	}
+
+	// In some cases, wrong-IP-family is only a problem when the secondary IP family
+	// isn't present at all.
+	if !dualStackSupported {
+		if len(s.Config.IPVS.ExcludeCIDRs) > 0 && !anyMatch(s.Config.IPVS.ExcludeCIDRs, okCIDR) {
+			errors = append(errors, fmt.Errorf("cluster is %s but ipvs.excludeCIDRs contains only %s addresses", clusterType, secondaryIPFamily))
+		}
+
+		if s.Config.HealthzBindAddress != "" {
```

Comment:
Does this handle "0.0.0.0:<port>" and ":<port>" correctly?  It reads as 4 but accepts on 6, too:

```
$ cat main.go 
package main

import (
	"fmt"
	"net"
	"os"
)

func main() {
	l, err := net.Listen("tcp", "0.0.0.0:9376")
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	l.Accept()
	fmt.Println("done")
}

$ go run main.go &
[1] 3125929

$ netcat -6 localhost 9376
done

[1]+  Done                    go run main.go

$ go run main.go &
[1] 3126055

$ netcat -4 localhost 9376
done

[1]+  Done                    go run main.go

```